crates/sc-subspace-block-relay/src/consensus.rs

```diff
@@ -725,25 +761,23 @@ where
     let backend = Arc::new(ConsensusClientBackend {
         transaction_pool: pool.clone(),
     });
-    let relay_client: ConsensusRelayClient<Block, Pool, _> = ConsensusRelayClient {
+    let relay_client: ConsensusRelayClient<Block, Pool, _> = ConsensusRelayClient::new(
         network,
-        protocol_name: SYNC_PROTOCOL.into(),
-        protocol: Arc::new(CompactBlockClient::new()),
+        SYNC_PROTOCOL.into(),
+        Arc::new(CompactBlockClient::new(RelayProtocol::CompactBlock)),
```

Comment:
Why is argument needed here? Could there be non-compact block relay protocol in compact block client?

Reply:
This allows assigning Ids to the protocols from a central place, without any clashes. This can be called anything else as well, name doesn't really matter

Reply:
Why do we need to assign protocol IDs? Why isn't it just an enum with incoming message. You match the variant and you know what protocol it is. We don't need to assign IDs as such and we don't need versions for each ID, at least the way I see it.

Reply:
On more thought, this protocol Id should not have enum either, changed it to be `u64`. Having it from central place is just extra precaution. When a new protocol is introduced, as long as it uses a different Identifier, we are good. So changed it so that the protocol implementation uses its own Id.

> You match the variant and you know what protocol it is. We don't need to assign IDs as such and we don't need versions for each ID, at least the way I see it.

This is the core issue (why we don't send enums across versions). I have a small test to demonstrate the problem:
```
#[cfg(test)]
mod tests {
    use codec::{Encode, Decode};

    #[derive(Debug, Encode, Decode, PartialEq, Eq)]
    enum Old {
        #[codec(index = 0)]
        E1,
    }

    #[derive(Debug, Encode, Decode, PartialEq, Eq)]
    enum New {
        #[codec(index = 1)]
        E2,
    }

    #[test]
    fn test_versioned() {
        let old = Old::E1;
        let encoded = old.encode();
        // This is fine.
        let decoded: Old = Decode::decode(&mut encoded.as_slice()).unwrap();
        assert_eq!(old, decoded);

        let new = New::E2;
        let encoded = new.encode();
        // This would fail:
        // Error { cause: None, desc: "Could not decode `Old`, variant doesn't exist" }
        assert!(<Old as Decode>::decode(&mut encoded.as_slice()).is_err());
    }
}
```

Decoding can fail either if the message is garbled, or if the encoded message is something the receiver does not understand. The versioning helps us catch a version mismatch *before* we try to decode, so we can differentiate it from a garbled message, and handle it appropriately.

One might ask: why not keep the older versions around for a while, until upgrade completes and rip them out after we are confident upgrade is complete? That makes the design complex: we need to keep multiple versions on server side, route requests appropriately, handle edge cases, etc. 

This design uses a simpler approach, where we only need to run a single protocol at any point in time. And fall back to serving full blocks when version mismatch is detected. Hope this clarifies

Reply:
> Decoding can fail either if the message is garbled, or if the encoded message is something the receiver does not understand. The versioning helps us catch a version mismatch before we try to decode, so we can differentiate it from a garbled message, and handle it appropriately.

The first byte of the encoded message is literally an enum variant number. It serves exactly the same purpose as version, except it is type safe and nicer to work with rather than anonymous `u64`. And if you can't decode the message, it matters very little why exactly that is the case, what matters is that node uses incompatible protocol and connection should be dropped right there.

> One might ask: why not keep the older versions around for a while, until upgrade completes and rip them out after we are confident upgrade is complete? That makes the design complex: we need to keep multiple versions on server side, route requests appropriately, handle edge cases, etc.

We definitely need to keep the older version around for a while or else we can't do graceful protocol upgrade.

> This design uses a simpler approach, where we only need to run a single protocol at any point in time. And fall back to serving full blocks when version mismatch is detected. Hope this clarifies

You can still do that with enums or we can just drop the connection. The roll-out will consist of enabling clients first and only later making new protocol the default, at which point clients should already be upgraded.

Reply:
> We definitely need to keep the older version around for a while or else we can't do graceful protocol upgrade.

Why? This change allows you to avoid all that complexity. Don't need to support multiple protocols at the same time, route the requests to the right handler, etc  (as explained above). Keep in mind, block relay is a simple request/response type stateless process. It is fine to not have the protocol path during the upgrade period IMO. Also, the sync path is still separate, where we send full blocks after the recent change

> You can still do that with enums or we can just drop the connection. The roll-out will consist of enabling clients first and only later making new protocol the default, at which point clients should already be upgraded.

Again, this avoids having to drop connections, deal with the edge cases. The upgrade would be fairly seamless

Reply:
> Why? This change allows you to avoid all that complexity. Don't need to support multiple protocols at the same time, route the requests to the right handler, etc (as explained above). Keep in mind, block relay is a simple request/response type stateless process. It is fine to not have the protocol path during the upgrade period IMO.

We need to go back to the reason why we want to have compact blocks in the first place, which is to maximize farmer bandwidth with useful data while avoiding increase of block propagation latency. In order to do that we want to broadcast small headers + as little extra data as possible rather than full blocks.

Now imagine we deploy an update where increasing number of nodes becomes incompatible. What does this mean? Well, this means that one or both of the following may happen:
* bandwidth requirements exceed available bandwidth on some nodes (they will have to spend up to 2x bandwidth comparing to before), causing them to fall behind the rest of the network
* block propagation latency will increase significantly, causing increased amount of forking, allowing for double spend attack on the network with less than 51% of storage or risk of permanent partitioning of the network

Both of those things are pretty bad in my opinion and are worth a bit of extra implementation complexity to avoid.

If designed with a simple multiplexer enum support for 2 protocols is not any more difficult than supporting any other number of protocols. Yes, there will be some extra logic for supporting second protocol comparing to just one, but it is not very large either.

Reply:
I understand why relay is needed, but question is: how long is the upgrade expected to last? I would like to see some concrete data instead of hand wavy arguments, before adding more complexity

> The roll-out will consist of enabling clients first and only later making new protocol the default, at which point clients should already be upgraded.

Every node acts as both relay client and server. Not sure what you mean here

TL;DR: I would rather not do the demux/routing on our side. This also requires creating a global enum of messages (across protocols), which breaks the protocol abstraction. Two approaches in my mind:
1. The scheme in this PR. I feel this is simple/clean.
2. Or change upstream to support multiple servers (or up to two at any point in time). This would also handle upgrades within same protocol gracefully (changes to message format, etc). I was looking at the upstream code, this looks feasible. The client side can still be as is today.

We can start a thread with upstream folks if needed.

Reply:
> I understand why relay is needed, but question is: how long is the upgrade expected to last? I would like to see some concrete data instead of hand wavy arguments, before adding more complexity

Judging from past experience from a few days to a few weeks. Either way I'm not ready to leave it up to chance.

> Every node acts as both relay client and server. Not sure what you mean here

I mean that initial roll-out will include support for new protocol, but nodes will stil make requests using old protocol. Second step in roll-out will be actual use of new protocol, but by then everyone is expected to be upgraded. Third step will be to remove support for old protocol as unnecessary and unused.

> I would rather not do the demux/routing on our side. This also requires creating a global enum of messages (across protocols), which breaks the protocol abstraction. 

It shouldn't if every protocol is self-contained and just receives request/response messages from somewhere. Then it doesn't matter if there is a multiplexer on top or not, protocol will work the same.

> The scheme in this PR. I feel this is simple/clean.

I don't find it simple because instead of protocol knowing almost or literally nothing about support for more than one protocol it does know and has more complex implementation thna before with no graceful upgrade path as far as I can tell.

> Or change upstream to support multiple servers (or up to two at any point in time). This would also handle upgrades within same protocol gracefully (changes to message format, etc). I was looking at the upstream code, this looks feasible. The client side can still be as is today.

I always considered the solution of using Substrate as a mid-term solution anyway. Eventual goal is to not use Substrate's networking and sync. Looks like you're suggesting to double-down on making their abstractions even more complex to handle our unusual (from their point of view) use case. This just increases our coupling with their networking/sync stacks and makes Substrate upgrades harder due to these stuff being not very needed upstream.

I really don't see how this is preferable to a top-level enum for messages.

Reply:
> Judging from past experience from a few days to a few weeks

I see, good to know. Quite surprising, as upgrade windows typically last 30 min - small number of hours (from operational point of view, it is preferable to have a uniform network as quickly as possible).

In light of this, it indeed makes sense to go through protocol path during upgrade

Reply:
> I see, good to know. Quite surprising, as upgrade windows typically last 30 min - small number of hours (from operational point of view, it is preferable to have a uniform network as quickly as possible).

Hm... I think we might be talking about different things. I meant it takes time for users to notice there is a new release and find time to actually install it. You can see distribution of releases in telemetry and compare to corresponding release dates.